executor/aggfuncs/func_group_concat.go

```diff
@@ -214,3 +221,298 @@ func (e *groupConcatDistinct) SetTruncated(t *int32) {
+			h.rows[0].buffer.Truncate(h.rows[0].buffer.Len() - int(debt))
+		} else {
+			h.currSize -= uint64(h.rows[0].buffer.Len()) + h.sepSize
+			heap.Pop(h)
+		}
+	}
+	return true
+}
+
+func (h *topNRows) reset() {
+	h.rows = h.rows[:0]
+	h.err = nil
+	h.currSize = 0
+}
+
+func (h *topNRows) concat(sep string, truncated bool) string {
+	buffer := new(bytes.Buffer)
+	sort.Sort(sort.Reverse(h))
+	for i, row := range h.rows {
+		if i != 0 {
+			buffer.WriteString(sep)
+		}
+		buffer.Write(row.buffer.Bytes())
+	}
+	if truncated && uint64(buffer.Len()) < h.limitSize {
+		// append the last separator, because the last separator may be truncated in tryToAdd.
+		buffer.WriteString(sep)
+		buffer.Truncate(int(h.limitSize))
+	}
+	return buffer.String()
+}
+
+type partialResult4GroupConcatOrder struct {
+	topN *topNRows
+}
+
+type groupConcatOrder struct {
+	baseGroupConcat4String
+}
+
+func (e *groupConcatOrder) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4GroupConcatOrder)(pr)
+	if p.topN.Len() == 0 {
+		chk.AppendNull(e.ordinal)
+		return nil
+	}
+	chk.AppendString(e.ordinal, p.topN.concat(e.sep, *e.truncated == 1))
+	return nil
+}
+
+func (e *groupConcatOrder) AllocPartialResult() PartialResult {
+	desc := make([]bool, len(e.byItems))
+	for i, byItem := range e.byItems {
+		desc[i] = byItem.Desc
+	}
+	p := &partialResult4GroupConcatOrder{
+		topN: &topNRows{
+			desc:      desc,
+			currSize:  0,
+			limitSize: e.maxLen,
+			sepSize:   uint64(len(e.sep)),
+		},
+	}
+	return PartialResult(p)
+}
+
+func (e *groupConcatOrder) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4GroupConcatOrder)(pr)
+	p.topN.reset()
+}
+
+func (e *groupConcatOrder) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) (err error) {
+	p := (*partialResult4GroupConcatOrder)(pr)
+	p.topN.sctx = sctx
+	v, isNull := "", false
+	for _, row := range rowsInGroup {
+		buffer := new(bytes.Buffer)
+		for _, arg := range e.args {
+			v, isNull, err = arg.EvalString(sctx, row)
+			if err != nil {
+				return err
+			}
+			if isNull {
+				break
+			}
+			buffer.WriteString(v)
+		}
+		if isNull {
+			continue
+		}
+		sortRow := sortRow{
+			buffer:  buffer,
+			byItems: make([]types.Datum, 0, len(e.byItems)),
+		}
+		for _, byItem := range e.byItems {
+			d, err := byItem.Expr.Eval(row)
+			if err != nil {
+				return err
+			}
+			sortRow.byItems = append(sortRow.byItems, d)
+		}
+		truncated := p.topN.tryToAdd(sortRow)
+		if p.topN.err != nil {
+			return p.topN.err
+		}
+		if truncated {
+			if err := e.handleTruncateError(sctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (e *groupConcatOrder) MergePartialResult(sctx sessionctx.Context, src, dst PartialResult) error {
+	return terror.ClassOptimizer.New(mysql.ErrInternal, mysql.MySQLErrName[mysql.ErrInternal]).GenWithStack("groupConcatOrder.MergePartialResult should not be called")
```

Comment:
Why return an error here? If we don't implement this method, we cannot use this aggregate function in `HashJoin`.

Reply:
See https://github.com/pingcap/tidb/pull/16591/files#diff-badfcd30d7596a08cd207b8e6ae778e6L1196, If `order by` exists, the parallel hash aggregation is forbidden.